ramips: add support for ASUS RT-AC1200-V2

therealsummoner opened the pull request:
Asus RT-AC1200-V2 wireless router

Hardware specifications:
SoC: MT7628DAN MIPS_24KEc@580MHz 2.4G-n 2x2
WiFi: MT7613BEN 5G-ac 160MHz 2x2
Switch: 4x100M built-in SoC
Flash: 16MB W25Q128JVSQ SPI-NOR
DRAM: 64MB built-in SoC

MAC addresses as verified by OEM firmware:
use            address      source
Lan/Wan/2G     *:60         factory 0x4 (label)
5G             *:64         factory 0x8000

Serial console: 57600,8n1

Installation:

Asus windows recovery tool:
 - install the Asus firmware restoration utility
 - unplug the router, hold the reset button while powering it on
 - release when the power LED flashes slowly
 - specify a static IP on your computer:
    IP address: 192.168.1.75
    Subnet mask 255.255.255.0
 - start the Asus firmware restoration utility, specify the factory image
    and press upload
 - do NOT power off the device after OpenWrt has booted until the LED flashing
 - after flashing OpenWrt, there will be first no 5GHz Wifi available probably,
    wait until blinking finishes and do a reboot

TFTP Recovery method:
 - set computer to a static ip, 192.168.1.75
 - connect computer to the LAN 1 port of the router
 - hold the reset button while powering on the router for a few seconds
 - send firmware image using a tftp client; i.e from linux:
    $ tftp
    tftp> binary
    tftp> connect 192.168.1.1
    tftp> put factory.bin
    tftp> quit
 - do NOT power off the device after OpenWrt has booted until the LED flashing
 - after flashing OpenWrt, there will be first no 5GHz Wifi available probably,
    wait until blinking finishes and do a reboot

Signed-off-by: Tamas Balogh <tamasbalogh@hotmail.com>

AuthorReflex commented:
for 5g led, setting 
```
	led {
			led-sources = <2>;
			led-active-low;
		};
```
in mt76 does not work? V2 has same chip as V1, so it must be working

thi-baut commented:
Hey thanks for this! Do you know if this would work with the ASUS RT-AC1200 (v1)?

AuthorReflex commented:
> Hey thanks for this! Do you know if this would work with the ASUS RT-AC1200 (v1)?

ASUS RT-AC1200 (v1) added in #5028. It's working.

therealsummoner commented:
Sadly this is a pretty much customized device by the OEM, at least in regard of LEDs for sure and thus it is possible it will be not accepted/supported due to relying on changes in two drivers it'd need for normal functionality.

therealsummoner commented:
> Since #5028 has mered, dts for v1 and v2 can be unified by make dts include file (dtsi)

Hi, can you check whether your device is ok as well with the dtsi changes?
E.g. if your LEDs are working and MAC addresses are correct?

w43322 commented:
Did some tests today, almost everything works. MAC addresses are correct, switch LEDs and 5G LED are OK, however the 2.4G LED, USB LED and power LED are not lighting up. Seems like the `leds` node didn't get overridden correctly.

w43322 commented:
```
asus,rt-ac1200|\
elecom,wrc-1167fs)
    wan_mac=$(mtd_get_mac_binary factory 0x22)
    label_mac=$wan_mac
    ;;
```
While you're at it, could u help me fix one thing? In `target/linux/ramips/mt76x8/base-files/etc/board.d/02_network` under the `ramips_setup_macs()` section, currently the v1 device shares its config with `elecom,wrc-1167fs`, pointing `label_mac` to `wan_mac`. Since `label_mac` is set in the dts, could u drop that line plz? (and in my previous commit it's incorrectly set anyways since the `label_mac` is the same as `lan` instead of `wan`)

On the v1 device, MAC addresses are as following (cross-checked with OEM firmware):
```
Lan/2G *:70 (label)
Wan/5G *:74
```
in EEPROM, the `*:70` MAC address appears at `0x4` and `0x22`, and the `*:74` MAC address appears at `0x28` and `0x8004` locations. 
So dropping that `label_mac=$wan_mac` line for the v1 device would suffice.
TIA

w43322 commented:
Also I've just noticed that the 2.4G signal on the device is very weak on rx. I think tx is fine, since the signal of connected devices are actually pretty good, but when operated in wwan mode and in channel analysis the 2.4G wifi can barely get anything (-80dbm for very close stations). Is it the same case in your device?

therealsummoner commented:
> ```
> asus,rt-ac1200|\
> elecom,wrc-1167fs)
>     wan_mac=$(mtd_get_mac_binary factory 0x22)
>     label_mac=$wan_mac
>     ;;
> ```
> 
> While you're at it, could u help me fix one thing? In `target/linux/ramips/mt76x8/base-files/etc/board.d/02_network` under the `ramips_setup_macs()` section, currently the v1 device shares its config with `elecom,wrc-1167fs`, pointing `label_mac` to `wan_mac`. Since `label_mac` is set in the dts, could u drop that line plz? (and in my previous commit it's incorrectly set anyways since the `label_mac` is the same as `lan` instead of `wan`)
> 
> On the v1 device, MAC addresses are as following (cross-checked with OEM firmware):
> 
> ```
> Lan/2G *:70 (label)
> Wan/5G *:74
> ```
> 
> in EEPROM, the `*:70` MAC address appears at `0x4` and `0x28`, and the `*:74` MAC address appears at `0x22` and `0x8004` locations. So dropping that `label_mac=$wan_mac` line for the v1 device would suffice. TIA

So I changed things, LEDs should work, please let me know if not.
Regarding MACs: I simply removed you from 02_network (just like in my case), because I don't want to interfere with the elecom device and secondly, OEM just takes the same MAC for wan, as the one for Lan and wlan2g. This is usually the case for Asus devices, i.e. it doesn't need to be defined at all.
Cheers

mans0n commented:
> OEM just takes the same MAC for wan, as the one for Lan and wlan2g

But @w43322 said otherwise? So v1 and v2 have different MAC address schemes? Weird...
And maybe we can also convert the elecom device to use dts approach for label mac.

w43322 commented:
> > OEM just takes the same MAC for wan, as the one for Lan and wlan2g
> 
> But @w43322 said otherwise? So v1 and v2 have different MAC address schemes? Weird... And maybe we can also convert the elecom device to use dts approach for label mac.

That appears to be the case. Here is what I've gathered from the OEM firmware using SSH:
```
# ifconfig -a
br0        Link encap:Ethernet  HWaddr XX:XX:XX:XX:XX:70
           inet addr:192.168.50.1  Bcast:192.168.50.255  Mask:255.255.255.0
           UP BROADCAST RUNNING ALLMULTI MULTICAST  MTU:1500  Metric:1
           RX packets:12148 errors:0 dropped:0 overruns:0 frame:0
           TX packets:21509 errors:0 dropped:0 overruns:0 carrier:0
           collisions:0 txqueuelen:0
           RX bytes:1718495 (1.6 MiB)  TX bytes:21179725 (20.1 MiB)

ra0        Link encap:Ethernet  HWaddr XX:XX:XX:XX:XX:70
           UP BROADCAST RUNNING ALLMULTI MULTICAST  MTU:1500  Metric:1
           RX packets:0 errors:0 dropped:0 overruns:0 frame:0
           TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
           collisions:0 txqueuelen:1000
           RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
           Interrupt:4

rai0       Link encap:Ethernet  HWaddr XX:XX:XX:XX:XX:74
           UP BROADCAST RUNNING ALLMULTI MULTICAST  MTU:1500  Metric:1
           RX packets:0 errors:0 dropped:0 overruns:0 frame:0
           TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
           collisions:0 txqueuelen:1000
           RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
           Interrupt:2

vlan2      Link encap:Ethernet  HWaddr XX:XX:XX:XX:XX:74
           inet addr:192.168.1.205  Bcast:192.168.1.255  Mask:255.255.255.0
           UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
           RX packets:263 errors:0 dropped:0 overruns:0 frame:0
           TX packets:295 errors:0 dropped:0 overruns:0 carrier:0
           collisions:0 txqueuelen:0
           RX bytes:130667 (127.6 KiB)  TX bytes:66596 (65.0 KiB)
```

w43322 commented:
All LEDs work after the change. 👍

w43322 commented:
Also I tried increasing the SPI frequency from 40MHz to 104MHz, but the difference is not that significant.
`Before:`
```
root@OpenWrt:~# time dd if=/dev/mtd3 of=/dev/null
32128+0 records in
32128+0 records out
real    0m 5.13s
user    0m 0.03s
sys     0m 4.87s
```
`After:`
```
root@OpenWrt:~# time dd if=/dev/mtd3 of=/dev/null
32128+0 records in
32128+0 records out
real    0m 3.92s
user    0m 0.00s
sys     0m 3.71s
```
Seems like the real clock speed is closer to 52MHz, but the device seems to be okay with it.

therealsummoner commented:
> > > OEM just takes the same MAC for wan, as the one for Lan and wlan2g
> > 
> > 
> > But @w43322 said otherwise? So v1 and v2 have different MAC address schemes? Weird... And maybe we can also convert the elecom device to use dts approach for label mac.
> 
> That appears to be the case. Here is what I've gathered from the OEM firmware using SSH:
> 
> ```
Yep, its different, I'll double check it however.

therealsummoner commented:
> Also I tried increasing the SPI frequency from 40MHz to 104MHz, but the difference is not that significant. `Before:`
> 
> ```
> root@OpenWrt:~# time dd if=/dev/mtd3 of=/dev/null
> 32128+0 records in
> 32128+0 records out
> real    0m 5.13s
> user    0m 0.03s
> sys     0m 4.87s
> ```
> 
> `After:`
> 
> ```
> root@OpenWrt:~# time dd if=/dev/mtd3 of=/dev/null
> 32128+0 records in
> 32128+0 records out
> real    0m 3.92s
> user    0m 0.00s
> sys     0m 3.71s
> ```
> 
> Seems like the real clock speed is closer to 52MHz, but the device seems to be okay with it.

The V2 device is setting it to 58MHz including fast-read mode; I was finding this directly from the SoC registers, so I can change it to 58, as I had it originally, but no more, to reflect the original operation. 104 is not available on V2, again as per SoC register setup.

w43322 commented:
> > Also I tried increasing the SPI frequency from 40MHz to 104MHz, but the difference is not that significant. `Before:`
> > ```
> > root@OpenWrt:~# time dd if=/dev/mtd3 of=/dev/null
> > 32128+0 records in
> > 32128+0 records out
> > real    0m 5.13s
> > user    0m 0.03s
> > sys     0m 4.87s
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `After:`
> > ```
> > root@OpenWrt:~# time dd if=/dev/mtd3 of=/dev/null
> > 32128+0 records in
> > 32128+0 records out
> > real    0m 3.92s
> > user    0m 0.00s
> > sys     0m 3.71s
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Seems like the real clock speed is closer to 52MHz, but the device seems to be okay with it.
> 
> The V2 device is setting it to 58MHz including fast-read mode; I was finding this directly from the SoC registers, so I can change it to 58, as I had it originally, but no more, to reflect the original operation. 104 is not available on V2, again as per SoC register setup.

Yes, I think setting it to 58MHz is OK.

therealsummoner commented:
> > > > OEM just takes the same MAC for wan, as the one for Lan and wlan2g
> > > 
> > > 
> > > But @w43322 said otherwise? So v1 and v2 have different MAC address schemes? Weird... And maybe we can also convert the elecom device to use dts approach for label mac.
> > 
> > 
> > That appears to be the case. Here is what I've gathered from the OEM firmware using SSH:
> 
> Yep, its different, I'll double check it however. As for the elecom device, I don't touch it in this PR.

Ok I was hooking up a cable to the WAN on the OEM FW and indeed is LAN, WAN and Wifi2G MACs the same.
Can you check which MACs you really see, so I mean not from ifconfig -a, but physically?

w43322 commented:
> 

I can confirm the WAN address is really the same as 5G wifi instead of LAN and 2.4G wifi, and that's actually how I got the information in the first place. I hooked it up to several different routers and got the same result.

therealsummoner commented:
> > 
> 
> I can confirm the WAN address is really the same as 5G wifi instead of LAN and 2.4G wifi, and that's actually how I got the information in the first place. I hooked it up to several different routers and got the same result.

Alright, I changed the SPI max clock and you should have now the correct Wan MAC as well.

w43322 commented:
> > > 
> > 
> > 
> > I can confirm the WAN address is really the same as 5G wifi instead of LAN and 2.4G wifi, and that's actually how I got the information in the first place. I hooked it up to several different routers and got the same result.
> 
> Alright, I changed the SPI max clock and you should have now the correct Wan MAC as well.

Yes, this should be ready now. 👍

therealsummoner committed:
ramips: add support for ASUS RT-AC1200-V2

Hardware specifications:
SoC: MT7628DAN MIPS_24KEc@580MHz 2.4G-n 2x2
WiFi: MT7613BEN 5G-ac 160MHz 2x2
Switch: 4x100M built-in SoC
Flash: 16MB W25Q128JVSQ SPI-NOR
DRAM: 64MB built-in SoC

MAC addresses as verified by OEM firmware:
use address source
Lan/Wan/2G *:60 factory 0x4 (label)
5G *:64 factory 0x8000

Serial console: 57600,8n1

Installation:

Asus windows recovery tool:

install the Asus firmware restoration utility
unplug the router, hold the reset button while powering it on
release when the power LED flashes slowly
specify a static IP on your computer:
IP address: 192.168.1.75
Subnet mask 255.255.255.0
start the Asus firmware restoration utility, specify the factory image
and press upload
do NOT power off the device after OpenWrt has booted until the LED flashing
after flashing OpenWrt, there will be first no 5GHz Wifi available probably,
wait until blinking finishes and do a reboot
TFTP Recovery method:

set computer to a static ip, 192.168.1.75
connect computer to the LAN 1 port of the router
hold the reset button while powering on the router for a few seconds
send firmware image using a tftp client; i.e from linux:
$ tftp
tftp> binary
tftp> connect 192.168.1.1
tftp> put factory.bin
tftp> quit
do NOT power off the device after OpenWrt has booted until the LED flashing
after flashing OpenWrt, there will be first no 5GHz Wifi available probably,
wait until blinking finishes and do a reboot

Signed-off-by: Tamas Balogh <tamasbalogh@hotmail.com>

hauke commented:
Thank you for your patch, I applied it to master in a4bf562aa71ad1e3dcffa392b79110d803a93f11.
